### ISSUE TITLE
Life bar(s) lag in reviewer: always one card behind the stats in the ankimon window

### ISSUE THREAD

**Issue by omfgroflmfaol** [Unlucky-Life/ankimon#61]
The inserted HP bar(s) always lag behind the ankimon window because `addHook("showQuestion", on_review_card)` is used for calculation/updating of the stats etc. whereas the hook `gui_hooks.reviewer_did_answer_card.append(update_life_bar)` is used for redrawing the reviewer life bar. Since `reviewer_did_answer_card` happens before `showQuestion`, there results a one card lag. To fix this, change the hook for `on_review_card` to

`gui_hooks.reviewer_did_answer_card.append(on_review_card)`

Because the hook `reviewer_did_answer_card` passes args, you need to make `on_review_cards` accept args to avoid an exception. This can be done by adding `(*args)` to its definition:

`def on_review_card(*args):`

Tested and works 👍

**Comment by face515** [comment]
Same happens to me, is this in config?

**Comment by CaptainDead** [comment]
Thanks for sharing a solution for the problem, I hope this helps unlucky to fix this issue faster🙏

**Comment by CaptainDead** [comment]
@Unlucky-Life  and does our health ever decrease? cuz then there is smwhat no point of our own health if it cannot decrease, so it would be great if our health also decreased, like when we press 'again' or 'hard' button, more decrease in our health when again pressed and less decrease in hard.

![image](https://github.com/Unlucky-Life/ankimon/assets/139457914/84185e86-99c4-43db-8c9e-b7bdd0f7f54b)

**Comment by omfgroflmfaol** [comment]
@face515 

> Same happens to me, is this in config?

no, the fixes I suggested need to be made inside of the` __init__.py` file inside the ankimons folder in the addins folder (you can find this file by going to the addins menu, selecting ankimon adding and clicking on "show files"). Even if you have no knowledge of coding you can still open this file with any text editor (though something specifically made for coding like visual studio code or sublime text is recommended for readability), make the changes I explained above, save the file and it will fix the issue (use search function to find the two lines that need changing in the code). Just remember make a backup copy of the file beforehand in case something breaks and you get crashes.

**Comment by omfgroflmfaol** [comment]
> @Unlucky-Life and does our health ever decrease? cuz then there is smwhat no point of our own health if it cannot decrease, so it would be great if our health also decreased, like when we press 'again' or 'hard' button, more decrease in our health when again pressed and less decrease in hard.

This is sadly not available yet, but it is already implemented and will be released with the next version according to the Update notes for upcoming version 1.27 (#52):

_New Option: Having to press again can damage own mainpokemon. If this mainpokemon is defeated - the wild pokemon flees and mainpokemon is revived (Note: a more complex concept of damaging of own pokemon in reviewer will be incorporated for the future.)_

**Comment by Unlucky-Life** [comment]
@omfgroflmfaol @CaptainDead Yes, right now the mainpokemon Hp was just for presentation to get a feeling of battle.
For the next update if you hit again or hard in a card round the next time your mainpokemon and the wild pokemon will get a change to attack. Meaning you will be damaged by "failing" a card !

**Comment by Unlucky-Life** [comment]
@omfgroflmfaol Thanks for the suggestion ! Ill test it out !

**Comment by CaptainDead** [comment]
Here's a suggestion
we get more damage for 'again' and less damage for 'hard', that would be kinda fun ig? and like give more damage to enemy on 'easy' and less on 'good'

**Comment by Unlucky-Life** [comment]
@CaptainDead Yes the more damage to an enemy is already integrated.
Basically:
Every round = Answering 2 Cards
If you press 2x good => the damage you are doing to enemy = multiplied by 1x
If you press 2x easy => the damage you are doing to an enemy = multiplied by 2x (basically a crit)
If you press 2x hard => the damage you are doing to an enemy = multiplied by 0.5x (basically you do half of the dmg)
if you press 2x again => the damage you are doing to an enemy = 0

**Comment by omfgroflmfaol** [comment]
@CaptainDead I think that's definitely possible! An issue with gamification in general is that you never want it to impede the learning. For some people, this scheme could incentivize them to not use the buttons hard and again and reward extensive use of the easy button. Then again it has always been one's own responsibility anyway, because

**Comment by omfgroflmfaol** [comment]
@CaptainDead I think that's definitely possible! An issue with gamification in general is that you never want it to impede the learning. For some people, any such penalization scheme could incentivize them use the buttons untruthfully, leading them avoid the again and hard buttons and to overuse the easy button. Then again, this has always been one's own responsibility anyway; pressing again has always been connected with the punishment of more reviews and easy has always been connected with reward through fewer reviews, even without any gamification elements.

I also like the idea of giving the option to additionally scale damage by answer speed:
- damage dealt after fast correct answer > after slow correct answer
- damage taken after fast incorrect answer < after slow incorrect answer 
The average answer time of a deck could help to individually calibrate this scale for each deck. Note how this creates homeostasis naturally: Together with the inverted scaling for correct and incorrect answers, this would naturally punish any tries to abuse the scaling mechanism. If you try always answering as quickly as possible, your risk to take larger damage in case of errors increases proportionally, while you also make it harder for yourself in the future to answer quicker than average. A safer approach taking time has lower risk but also yields increasingly lower damage, making it unviable (as well as boring) to always take too long.

This could of course also be combined with the multipliers already in place for the answer keys.

Although I must admit what I would like to see the most is an enemy AI being implemented that attacks regularly with the actual pokemon's specific move set - alongside the ability to choose attacks as a player (maybe via a popup dialog where the move can be chosen with the keys 1,2,3,4 as well). This would make battles more interesting strategically, although the learning would frequently be interrupted by gameplay decisions. Personally, I would love it - this way it would feel more like I'm learning cards by playing pokemon than like playing pokemon by learning cards, if this makes sense 😁  

I think in any case there should be different options to choose from: Random attacks w/o damage for maximum non-intrusive learning experience (your just also collecting and leveling), then also with taking damage (adding the challenge of needing to stay alive by being accurate), then also with choosing specific attacks and responding to specific attacks of the enemy pokemon via the choice of a pokemon from your team (adding strategy as a purely fun element)

**Comment by Unlucky-Life** [comment]
@omfgroflmfaol @CaptainDead The idea of actually choosing an attack, was actually a idea i implemented in the first version of ankimon => where in the ankimon window, the 4 attacks would appear that you could choose - The issue being that exactly the learning "speed" would be interrupted.

So basically - i would already have the code, which wouldnt take that long to reintegrate. But make it a more smooth option, was the issue. Aswell as creating a enemy ai could be quite complicated, due to Anki not allowing outside anki libraries.


I think i would integrate in the future. Right now im thinking about the option to have extra option to fight arena battles. Which would only be possible if you have learned your cards. => Basically you could be able to play small pokemon levels / events in your browser through javascript and html.

**Comment by omfgroflmfaol** [comment]
That's a great idea! I like that even more! To have such a big reward waiting at the end of a completed learning session seems very motivating. I think this would work very well :)

**Comment by omfgroflmfaol** [comment]
I noticed that `update_life_bar` should also be called after a new pokemon appears (after choosing whether to defeat or keep)

**Comment by Unlucky-Life** [comment]
@omfgroflmfaol What do you mean with that exactly ?

**Comment by Unlucky-Life** [comment]
@omfgroflmfaol because the issue being: The ankimon window buttons defeat and catch pokemon are both also calling the function (new_pokemon) and this seems to not work, because update_life_bar needs the reviewer html content to modify it

**Comment by omfgroflmfaol** [comment]
Hmm yes I overlooked this. After fiddling around a bit, I made it work by just appending the following lines to the `new_pokemon` function:

`from aqt import mw`
`class Container(object):`
`    pass`
`reviewer = Container()`
`reviewer.web = mw.reviewer.web`
`update_life_bar(reviewer, 0, 0)`

Seems a bit dirty, but the idea is to just read out the mw.reviewer.web contents and place them where update_life_bar would expect them. 0, 0 because it expects three args. 

I tested it and the overlay now updates immediately after catching or defeating. Haven't run into any issues so far :)

By the way, is there an easier way for me to send you code snippets than through here? I'm completely new to GitHub, but should I try using pull requests in the future?
